### PR TITLE
[Jobs] Make sure JobConfig specified in the init script is not being overridden by the SupervisorActor

### PR DESCRIPTION
**Pull request by alexeykudinkin**
## Why are these changes needed?

NOTE: This is a revert of the [revert](https://github.com/ray-project/ray/pull/38655), addressing broken tests

Currently when executing the job submitted via Ray Job submission API, `SupervisorActor` will be overriding whole of the `JobConfig` making it impossible to specify additional params like `JobConfig.code_search_path` for ex.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -1416,16 +1419,26 @@ def init(
...
+        )
+        # NOTE: We always prefer runtime_env injected via RAY_JOB_CONFIG_JSON_ENV_VAR,
+        #       as compared to via ray.init(runtime_env=...) to make sure runtime_env
+        #       specified via job submission API takes precedence
+        runtime_env = injected_job_config.runtime_env
 
         if ray_constants.RAY_RUNTIME_ENV_HOOK in os.environ and not _skip_env_hook:
             runtime_env = _load_class(os.environ[ray_constants.RAY_RUNTIME_ENV_HOOK])(
-                job_config.runtime_env
+                runtime_env
```
Comment (alexeykudinkin) [inline python/ray/_private/worker.py:1435]: This line is the only change as compared to https://github.com/ray-project/ray/pull/38655

**Review by architkulkarni — APPROVED** [review]

**Comment by architkulkarni** [comment]
@alexeykudinkin potentially related test failure  https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8245 (or https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8229 or https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8218, not sure which one is the root cause)

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
Test failure: https://buildkite.com/ray-project/oss-ci-build-pr/builds/32913#018a198f-5512-49d0-9b6b-f4d11026424e/3356-8768

**Comment by alexeykudinkin** [comment]
Thanks for reviewing @edoakes @architkulkarni!

Yeah, i figured that new test that i've added was messing with our Ray cluster sharing in other tests and fixed that, but when i've tried to run other tests locally it turned out that there quite a few more tests that were failing in `test_job_manager`: basically all of the tests doing `ray start --head` seem to have been broken for quite some time (b/c these are not provided `--address` flag which seemed to be expected by `call_ray_start` fixture (and it's broken on master as well).

Regardless, i've fixed that one as well, and now it should be working fine.

File: python/ray/tests/conftest.py [hunk 2]
```diff
@@ -589,11 +589,16 @@ def call_ray_start_context(request):
     except Exception as e:
         print(type(e), e)
         raise
+
     # Get the redis address from the output.
     redis_substring_prefix = "--address='"
-    address_location = out.find(redis_substring_prefix) + len(redis_substring_prefix)
-    address = out[address_location:]
-    address = address.split("'")[0]
+    idx = out.find(redis_substring_prefix)
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply python/ray/tests/conftest.py:595]: Ah I think the discussion here might be related https://github.com/ray-project/ray/pull/35226. It got reverted for breaking other tests, sorry for not following through and unreverting it.

**Comment by alexeykudinkin** [comment]
Checked out test failures and all seems unrelated.

@edoakes @architkulkarni can you folks help merging this in?

**Comment by architkulkarni** [comment]
@alexeykudinkin https://buildkite.com/ray-project/premerge/builds/2812#018a1e80-6f06-4fa9-8fd0-9a08f35221ca/166-406 related?

**Comment by alexeykudinkin** [comment]
@architkulkarni fixed linter's failures.

Current failures don't seem to be related.

**Comment by architkulkarni** [comment]
Failures
- Windows tests:test_streaming_generator::test_status_task_events_metrics is not flaky on master. But it seems unrelated
- rllib:TestLearnerGroupAsyncUpdate flaky on master and unrelated
- tune:test_commands flaky on master and unrelated
- chaos network delay test flaky on master and unrelated

**Pull request merged**